src/ddsrt/src/library/windows/library.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <stdio.h>
+#include <assert.h>
+#include <string.h>
+#include "dds/ddsrt/heap.h"
+#include "dds/ddsrt/types.h"
+#include "dds/ddsrt/string.h"
+#include "dds/ddsrt/library.h"
+
+
+dds_retcode_t
+ddsrt_dlopen(
+    const char *name,
+    bool translate,
+    ddsrt_lib *handle)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+
+    if ( name == NULL || handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+
+    } else {
+        *handle = NULL;
+        GetLastError(); /* Clear any existing error */
+        if ((translate) && (strrchr(name, '/') == NULL )
+                        && (strrchr(name, '\\') == NULL )) {
+            /* Add suffix to the name and try to open. */
+            static const char suffix[] = ".dll";
+            size_t len = strlen(name) + sizeof(suffix);
+            char* libName = ddsrt_malloc(len);
+            sprintf_s(libName, len, "%s%s", name, suffix);
+            *handle = LoadLibrary(libName);
+            ddsrt_free(libName);
+        }
+
+        if (*handle == NULL && retcode == DDS_RETCODE_OK) {
+            /* Name contains a path,
+             * (auto)translate is disabled or
+             * LoadLibrary on translated name failed. */
+            *handle = LoadLibrary(name);
+        }
+
+        if (*handle != NULL) {
+            retcode = DDS_RETCODE_OK;
+        } else {
+            retcode = DDS_RETCODE_ERROR;
+        }
+    }
+
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlclose(
+    ddsrt_lib handle)
+{
+    dds_retcode_t retcode;
+
+    if( handle == NULL ){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    }
+    else{
+        retcode = (FreeLibrary(handle) == 0) ? DDS_RETCODE_ERROR : DDS_RETCODE_OK;
+    }
+
+    return retcode;
+}
+
+dds_retcode_t
+ddsrt_dlsym(
+    ddsrt_lib handle,
+    const char *symbol,
+    void **address)
+{
+    dds_retcode_t retcode = DDS_RETCODE_OK;
+
+    if ( symbol == NULL || handle == NULL || address == NULL){
+        retcode = DDS_RETCODE_BAD_PARAMETER;
+    } else {
+        *address  =  GetProcAddress(handle, symbol);
+        if( *address == NULL ){
+            retcode = DDS_RETCODE_ERROR;
+        }
+    }
+
+    return retcode;
+}
+
+
+dds_retcode_t
+ddsrt_dlerror(char *buf, size_t buflen)
```

Comment:
This doesn't do the same thing as the `ddsrt_dlerror` implementation in the POSIX layer above. `GetLastError` is used by much more of the api. It's better to use a thread-local variable to store the error when it occurred in one of the functions. Of course just dropping the `ddsrt_dlerror` call entirely is also a possibility. @eboasson do you have a preference here?

Reply:
Yes, you're right. However, GetLastError stores the error message for each thread. So, if the user calls the ddsrt_dlerror just after calling the other library functions, he/she will get the error message properly.

Reply:
That's not what the documentation blurb says ("Get the most recent library related error.") You're now suggesting that it should be extended with "and you may not have intervening calls to any Win32 functions on the same thread". While I agree that commonly one would expect dlerror to be called immediately, there is something to be said in favour of implemting the interface properly.

Reply:
I've updated the description